### ISSUE TITLE
Free strings passed into sherpa-onnx

### ISSUE THREAD

**Issue by thewh1teagle** [thewh1teagle/sherpa-rs#24]
Sometimes sherpa-onnx require string argument. so we use the macro `cstr!(...).into_raw()` and tell Rust to forget about the string so it won't be freed. However It's not clear where it's cleared and if we should clean it in some time.

**Comment by csukuangfj** [comment]
you can clear it after you have used it to create a recognizer.

That is, once you have a recognizer, you can free it.

**Comment by ghjdegithub** [comment]
> Sometimes sherpa-onnx require string argument. so we use the macro `cstr!(...).into_raw()` and tell Rust to forget about the string so it won't be freed. However It's not clear where it's cleared and if we should clean it in some time.

Then you're exposing methods that are not safe and have memory leaks.
It may be necessary to wrap a secure method or a secure structure.

**Comment by csukuangfj** [comment]
To give you a concrete example, 

https://github.com/thewh1teagle/sherpa-rs/blob/ae460f525ddcf7c9a37e0e3f609aa9c8850f2793/src/whisper.rs#L116

You can free all strings used by `config` after the above line.

**Comment by ghjdegithub** [comment]
cstr! = std::ffi::CString::new($s).expect("Failed to create CString").
Your macro I read is transferring ownership out to ffi's code, it's logical that ffi's code releases it, are you trying to save the pointer to this string and release it yourself, or maybe ffi will release it itself?

**Comment by csukuangfj** [comment]
The C code won't free any pointer that is created from the Rust code.

**Comment by csukuangfj** [comment]
>  or maybe ffi will release it itself?

No, it won't.

**Comment by ghjdegithub** [comment]
> The C code won't free any pointer that is created from the Rust code.C 代码不会释放从 Rust 代码创建的任何指针。

That would require rust to release the String itself.
I read the documentation for CString, and the documentation for into_raw says that it needs to be reconstructed using CString::from_raw (the pointer returned by into_raw) to release it correctly.

**Comment by thewh1teagle** [comment]
> I read the documentation for CString, and the documentation for into_raw says that it needs to be reconstructed using CString::from_raw (the pointer returned by into_raw) to release it correctly.

Exactly. I created new macro for doing it. It's a bit less comfortable now because we need to create pointers and free them but that looks the valid way.

```rust
#[macro_export]
macro_rules! free_cstr {
    ($ptr:expr) => {
        if !$ptr.is_null() {
            // Reclaim the CString to ensure it gets dropped and its memory is freed
            unsafe {
                let _ = std::ffi::CString::from_raw($ptr as *mut i8);
            }
        }
    };
}

let provider_ptr = cstr!(provider.clone()).into_raw();
let config = { ... };
// Create speaker diarization
let sd = unsafe { sherpa_rs_sys::SherpaOnnxCreateOfflineSpeakerDiarization(&config) };
free_cstr!(embedding_model_ptr);
```

**Comment by ghjdegithub** [comment]
> ```rust
> sherpa_rs_sys
> ```

It's really not rusty to do this. 😭

**Comment by thewh1teagle** [comment]
> > ```rust
> > sherpa_rs_sys
> > ```
> 
> It's really not rusty to do this. 😭

What do you mean? the use of the macros to create pointers and free them?
Do you have better idea?

**Comment by ghjdegithub** [comment]
> > > ```rust
> > > sherpa_rs_sys
> > > ```
> > 
> > 
> > It's really not rusty to do this. 😭
> 
> What do you mean? the use of the macros to create pointers and free them? Do you have better idea?

I mean to construct a smart pointer struct, struct inside a ptr is the pointer returned by into_raw, drop will release this into_raw returned pointer

**Comment by thewh1teagle** [comment]
> I mean to construct a smart pointer struct, struct inside a ptr is the pointer returned by into_raw, drop will release this into_raw returned pointer

That sounds better. I'll wee if I can make sure it won't will drop before the function ends. Maybe that the default behavior

**Comment by ghjdegithub** [comment]
> > I mean to construct a smart pointer struct, struct inside a ptr is the pointer returned by into_raw, drop will release this into_raw returned pointer
> 
> 
> 
> That sounds better. I'll wee if I can make sure it won't will drop before the function ends. Maybe that the default behavior 

Yes, construct a smart pointer that contains the pointer returned by into_raw, and then the smart pointer will automatically release the pointer returned by into_raw when it leaves scope. The key is how to use this smart pointer, creating and passing it and when to automatically release it, all of which requires complex design.

**Comment by thewh1teagle** [comment]
@ghjdegithub 

Thanks. I thought for some reason that rust will detect that it goes out of the scope just when I call with it. I see that the scope ends when the function ends

**Comment by thewh1teagle** [comment]
Fixed in 0.5.0 with @ghjdegithub suggestion.
Thank you all!
